README
Gnome2::GConf version 1.045
===========================

This module allows you to use the GConf configuration system in order to
store/retrieve the configuration of an application.  GConf is a powerful
configuration manager based on a user daemon that handles a set of key and
value pairs, and notifies any changes of the value to every program that
monitors those keys.  GConf is used by GNOME 2.x.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install


To avoid installing to a system directory, since this is a beta release,
you can change the installation prefix at Makefile.PL time with

   perl Makefile.PL PREFIX=/some/other/place

This will install the module to the subdirectory lib/perl5 under the given
prefix.  If this is not already in perl's include path, you'll need to tell
perl how to get to this library directory so you can use it; there are three
ways:

  in your environment (the easiest):
    PERL5LIB=/some/other/place/lib/perl5/site_perl:$PERL5LIB

  on the perl command line:
    perl -I /some/other/place/lib/perl5/site_perl yourscript

  in the code of your perl script:
    use lib '/some/other/place/lib/perl5/site_perl';


DEPENDENCIES

This module requires these other modules and libraries:

  perl >= 5.8.0
  Glib >= 1.120 (perl module and requisite C libraries)
  libgconf2 >= 2.0.0 (C library)

To build from source, you also need:

  ExtUtils::Depends >= 0.2
  ExtUtils::PkgConfig >= 1.03

COPYRIGHT AND LICENSE

Copyright (C) 2003-2006 by Emmanuele Bassi <emmanuele.bassi@iol.it>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, see
<https://www.gnu.org/licenses/>.


NEWS
Overview of changes since 1.044
===============================

* Added 'doap' file with a lowercased filename for use by GNOME GitLab
* xs/GConfSchema.xs: applied patch from Debian project; closes RT#95015
* xs/GConf2.xs,xs/GConfClient.xs: Applied docpatch from Debian project; closes
  RT#95014
* META.yml: added 'configure_requires' key, this fixes automated installs via
  cpanminus (RT#129419)
* xs/GConf2.xs: fixed grammar error (extra 'that')
* Updated FSF address in all files that have it

Overview of changes since 1.043
===============================

* Make it possible for other bindings to use symbols from Gnome2::GConf's API.

Overview of changes since 1.042
===============================

* Treat Gnome2::GConf::Schema objects as fundamental types
  in Gnome2::GConf::Value.
* Fix usage of Gnome2::GConf::Client getters for keys of type
  'schema' (#361103, Laurent Simonneau)
* Fix a leak inside Gnome2::GConf::Schema.

Overview of changes since 1.041
===============================

* Implement GConf::Client::recursive_unset() to recursively
  unset a directory in the database. (#361094, based on a patch
  by Laurent Simonneau)

Overview of changes since 1.040
===============================

* Fix a leak in GConfValue. (#358294, Santeri Paavolainen)

Overview of changes since 1.031
===============================

* Fix loaded flags on Darwin. [Daniel Macks]

Overview of changes since 1.020
===============================

* The "An Address Indicates Where It Is" release.
* Wrap Gnome2::GConf::Engine::get_for_addresses. [Emmanuele]

Overview of changes since 1.013
===============================

* The "Differential Engine" release
* Fix Gnome2::GConf::Engine reference mechanism. [Emmanuele] 
* Use Glib::CodeGen from the stable Glib release. [Emmanuele]

Overview of changes since 1.012
===============================

* The "Breaking up is hard to do" release.
* Implemented the new error handling system. [Emmanuele]
* Changed Gnome2::GConf::Client::all_entries semantics. [Emmanuele]
* Documentation fixes. [Emmanuele]

Overview of changes since 1.011
===============================

* Finished the GConfEngine bindings. [Emmanuele]

Overview of changes since 1.010
===============================

* Require GConf 2.0.0. [Emmanuele]

Overview of changes since 1.000
===============================

* Wrapped GConfEngine. [Emmanuele]
* Documentation fixes. [Emmanuele]
* Require Glib 1.070. [Emmanuele]

Overview of changes since 0.94
==============================

* New test suite for GConfClient. [Emmanuele]
* Adapted version information methods to the current policy. [Toersten]
* Updated documentation. [Emmanuele]
* Require Glib 1.040. [Emmanuele]

Overview of changes since 0.93
==============================

* Fixed a stupid bug in GConfValue [Emmanuele]

Overview of changes since 0.92

* Split-up API documentation for GConfValue, GConfEntry,
  GConfChangeSet and GConfSchema. [Emmanuele]
* API frozen [Emmanuele]
* Require Glib 1.038 (for API freeze state) [Emmanuele]

Overview of changes since 0.91
==============================

* Require Glib 1.037 [Emmanuele]
* Use the new Glib::Error exception handling. [Emmanuele]
* Updated documentation. [muppet]

Overview of changes since 0.46
==============================

* Require ExtUtils::PkgConfig 1.03
* Require Glib 1.020
* Remove runtime dependency on Gtk2; Gtk2's Gtk2::CodeGen is a
  build-only requirement.  [muppet]
* Build system updates.  [muppet]
* Updated documentation.  [Emmanuele]
* Added some test methods that apps should not use.  [Emmanuele]

[started NEWS]


GConf.pm
#
# Copyright (c) 2003-2006 by Emmanuele Bassi (see the file AUTHORS)
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Library General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public
# License along with this library; if not, see
# <https://www.gnu.org/licenses/>.

package Gnome2::GConf;

use 5.008;
use strict;
use warnings;

use Glib;

require DynaLoader;

our @ISA = qw(DynaLoader);

our $VERSION = '1.045';

sub dl_load_flags { $^O eq 'darwin' ? 0x00 : 0x01 }

Gnome2::GConf->bootstrap ($VERSION);


# Preloaded methods go here

package Gnome2::GConf::Value;

use overload
	'==' => sub { Gnome2::GConf::Value::compare($_[0], $_[1]) },
	fallback => 1;

use overload
	'""' => sub { Gnome2::GConf::Value::to_string($_[0]) },
	fallback => 1;

package Gnome2::GConf::Client;
use Carp;

sub get_list
{
	my $self = shift;	# the object
	my $key  = shift;
	my $check_error = shift || 1;

	my $val = $self->get($key, $check_error);
	return $val->{value};
}

sub get_pair
{
	my $self = shift;	# the object
	my $key  = shift;
	my $check_error = shift || 1;

	my $val = $self->get($key, $check_error);
	carp "$key is not bound to a pair" if not $val->{type} eq 'pair';

	return ($val->{car}, $val->{cdr});
}

sub set_list
{
	my $self = shift;	# the object
	my $key  = shift;
	my $type = shift;
	my $list = shift;
	my $check_error = shift || 1;

	$self->set($key, { type => $type, value => $list }, $check_error);
}

sub set_pair
{
	my $self = shift;	# the object
	my $key  = shift;
	my $car  = shift;
	my $cdr  = shift;
	my $check_error = shift || 1;

	$self->set($key, {
			type	=> 'pair',
			car		=> $car,
			cdr		=> $cdr,
		}, $check_error);
}

package Gnome2::GConf;

1;
__END__

=head1 NAME

Gnome2::GConf - Perl wrappers for the GConf configuration engine.

=head1 SYNOPSIS

  use Gnome2::GConf;

  my $client = Gnome2::GConf::Client->get_default;
  my $app_key = "/apps/myapp/mykey";
  
  $client->add_dir($app_key, 'preload-none');
  
  # add a notify for the key
  my $notify_id = $client->notify_add($app_key, sub {
      my ($client, $cnxn_id, $entry) = @_;
      return unless $entry->{value};
		
      if ($entry->{value}->{type} eq 'string') {
        printf "key '%s' changed to '%s'\n",
	       $entry->{key},
	       $entry->{value}->{value};
      }
    });
  
  my $string = $client->get_string($app_key);
  $string = 'some string' unless $string;
  
  $client->set($app_key, { type => 'string', data => $string });
  
  # set a schema for the key
  $client->set_schema ($app_key,
      {
        type => 'string',
	locale => 'C',
	short_desc => 'Some key.',
	long_desc => 'This key does something.',
	owner => 'some_program'
      });
  
  # remove the notification callback
  $client->notify_remove($notify_id);

=head1 ABSTRACT

Perl bindings to the 2.2 series of the GConf configuration engine
libraries, for use with gtk2-perl.

=head1 DESCRIPTION

This module allows you to use the GConf configuration system in order
to store/retrieve the configuration of an application.  The GConf
system is a powerful configuration manager based on a user daemon that
handles a set of key and value pairs, and notifies any changes of the
value to every program that monitors those keys.  GConf is used
by GNOME 2.x.

To discuss gtk2-perl, ask questions and flame/praise the authors,
join gtk-perl-list@gnome.org at lists.gnome.org.

Find out more about Gnome at http://www.gnome.org.

=head1 DIFFERENT DATA TYPES

Some opaque data types in GConf are not registered inside the Glib type
system.  Thus, they have been implemented in a more perlish way, when
possible, for the sake of coherency and following the principle of
least surprise for the perl developer.  These changes try to preserve
semantics, to add syntactic sugar and to remove the need for accessor
methods.

=over

=item GConfEntry

See L<Gnome2::GConf::Entry>

=item GConfValue

See L<Gnome2::GConf::Value>

=item GConfChangeSet

See L<Gnome2::GConf::ChangeSet>

=item GConfSchema

See L<Gnome2::GConf::Schema>

=back

=head1 DIFFERENT CALL SIGNATURES

Reflecting the changes operated for the data types, some methods that use
those type have had the call signature modified.

=over

=item GConfNotifyFunc

In C, the function passed to C<Gnome2::GConf::notify_add> must have the
following signature:
	
	void (GConfNotifyFunc *) (GConfClient * client,
	                          guint cnxn_id,
	                          GConfEntry * entry);

Where C<GConfEntry> is a container for the key/value pair.  Since in
perl there's no C<GConfEntry> (see above), the C<entry> parameter is
an hashref.

=item GConfClient::get

=item GConfClient::set

In C, these accessor methods return/use a C<GConfValue>.  In perl, they
return/use an hashref.  See L<Gnome2::GConf::Value>

=item GConfClient::get_list

=item GConfClient::set_list

These accessor methods use a string for setting the type of the lists
(lists may have values of only B<one> type), and an arrayref containing
the values.

=item GConfClient::get_pair

=item GConfClient::set_pair

These accessor methods use two hashref (representing C<GConfValue>s)
for the C<car> and the C<cdr> parameters.

=item GConfClient::get_schema

=item GConfClient::set_schema

Similarly to the get/set pair above, these two methods return/use an
hashref. See L<Gnome2::GConf::Schema>.

=item GConfClient::commit_change_set

In C, this method return a boolean value (TRUE on success, FALSE on
failure). On user request (using the boolean parameter
C<remove_committed>), it also returns the C<GConfChangeSet>, pruned of
the successfully committed keys.  In perl, this method returns a
boolean value both in scalar context or if the user sets to FALSE the
C<remove_committed> parameter; in array context or if the user requests
the uncommitted keys, returns both the return value and the pruned
C<GConfChangeSet>.

=back

=head1 SEE ALSO

L<perl>(1), L<Glib>(3pm).

=head1 AUTHOR

Emmanuele Bassi E<lt>ebassi@gmail.comE<gt>

gtk2-perl created by the gtk2-perl team: http://gtk2-perl.sf.net

=head1 COPYRIGHT AND LICENSE

Copyright 2003-2006 by Emmanuele Bassi

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, see
<https://www.gnu.org/licenses/>.

=cut


examples/basic-gconf-app.pl
#!/usr/bin/perl

# This program demonstrates how to use GConf.  The key thing is that
# the main window and the prefs dialog have NO KNOWLEDGE of one
# another as far as configuration values are concerned; they don't
# even have to be in the same process. That is, the GConfClient acts
# as the data "model" for configuration information; the main
# application is a "view" of the model; and the prefs dialog is a
# "controller."
#
# You can tell if your application has done this correctly by
# using "gconftool" instead of your preferences dialog to set
# preferences. For example:
# 
# gconftool --type=string --set /apps/basic-gconf-app/foo "My string"
# 
# If that doesn't work every bit as well as setting the value
# via the prefs dialog, then you aren't doing things right. ;-)
#
#
# If you really want to be mean to your app, make it survive
# this:
# 
# gconftool --break-key /apps/basic-gconf-app/foo
# 
# Remember, the GConf database is just like an external file or
# the network - it may have bogus values in it. GConf admin
# tools will let people put in whatever they can think of.
# 
# GConf does guarantee that string values will be valid UTF-8, for
# convenience.
# 

# Throughout, this program is letting GConfClient use its default
# error handlers rather than checking for errors or attaching custom
# handlers to the "unreturned_error" signal. Thus the last arg to
# GConfClient functions is None.
#

# Special mention of an idiom often used in GTK+ apps that does
# not work right with GConf but may appear to at first:
#
# i_am_changing_value = gtk.TRUE
# change_value (value)
# i_am_changing_value = gtk.FALSE
# 
# This breaks for several reasons: notification of changes
# may be asynchronous, you may get notifications that are not
# caused by change_value () while change_value () is running,
# since GConf will enter the main loop, and also if you need
# this code to work you are probably going to have issues
# when someone other than yourself sets the value.
# 
# A robust solution in this case is often to compare the old
# and new values to see if they've really changed, thus avoiding
# whatever loop you were trying to avoid.
#

# The code is a direct mapping (with some perlisms) of the C code; where the
# code diverge, I placed a comment. (ebassi)

use strict;
use warnings;

use Glib qw/TRUE FALSE/;
use Gtk2 '-init';
use Gnome2::GConf;

our $client = Gnome2::GConf::Client->get_default;

# Tell GConfClient that we're interested in the given directory.
# This means GConfClient will receive notification of changes
# to this directory, and cache keys under this directory.
# So _don't_ add "/" or something silly like that or you'll end
# up with a copy of the whole GConf database. ;-)
#
# We use 'preload_none' to avoid loading all config keys on
# startup. If your app pretty much reads all config keys
# on startup, then preloading the cache may make sense.
   
$client->add_dir ("/apps/basic-gconf-app", 'preload-none');

our $main_window = create_main_window ($client);
$main_window->show_all;

Gtk2->main;

$client->remove_dir ("/apps/basic-gconf-app");

0;

sub create_main_window
{
	my $client = shift;
	
	my $w = Gtk2::Window->new('toplevel');
	$w->set_title('basic-gconf-app Main Window');
	
	my $vbox = Gtk2::VBox->new(FALSE, 5);
	$vbox->set_border_width(5);

	$w->add($vbox);

	my $config;
	
	# Create labels that we can "configure"
	$config = create_configurable_widget ($client, "/apps/basic-gconf-app/foo");
	$vbox->pack_start($config, TRUE, TRUE, 0);

	$config = create_configurable_widget ($client, "/apps/basic-gconf-app/bar");
	$vbox->pack_start($config, TRUE, TRUE, 0);

	$config = create_configurable_widget ($client, "/apps/basic-gconf-app/baz");
	$vbox->pack_start($config, TRUE, TRUE, 0);

	$config = create_configurable_widget ($client, "/apps/basic-gconf-app/blah");
	$vbox->pack_start($config, TRUE, TRUE, 0);

	$w->signal_connect(destroy => sub { Gtk2->main_quit });
	$w->{client} = $client;

	my $prefs = Gtk2::Button->new("Prefs");
	$vbox->pack_end($prefs, FALSE, FALSE, 0);
	$prefs->signal_connect(clicked => sub {
			my $button = shift;
			my $main_window = shift;

			my $prefs_dialog = $main_window->{prefs};
			if (not $prefs_dialog)
			{
				my $client = $main_window->{client};
				$prefs_dialog = create_prefs_dialog ($main_window, $client);
				$main_window->{prefs} = $prefs_dialog;
				
				$prefs_dialog->signal_connect(
						destroy => \&prefs_dialog_destroyed,
						$main_window);

				$prefs_dialog->show_all;
			}
			else
			{	
				# show existing dialog
				$prefs_dialog->present;
			}
		}, $w);
		
	return $w;
}

# Create a GtkLabel inside a frame, that we can "configure"
# (the label displays the value of the config key).
sub create_configurable_widget
{
	use Data::Dumper;
	
	my $client = shift;
	my $config_key = shift;

	my $frame = Gtk2::Frame->new($config_key);
	my $label = Gtk2::Label->new;
	$frame->add($label);

	my $s = $client->get_string($config_key);
	$label->set_text($s) if $s;

	my $notify_id = $client->notify_add($config_key, sub {
			# Notification callback for our label widgets that
			# monitor the current value of a gconf key. i.e.
			# we are conceptually "configuring" the label widgets
			my ($client, $cnxn_id, $entry, $label) = @_;
			return unless $entry;
			
			# Note that value can be undef (unset) or it can have
			# the wrong type! Need to check that to survive
			# gconftool --break-key
			unless ($entry->{value})
			{
				$label->set_text('');
			}
			elsif ($entry->{value}->{type} eq 'string')
			{
				warn(sprintf("got: %s\n", $entry->{value}));

				$label->set_text($entry->{value}->{value});
			}
			else
			{
				$label->set_text('!type error!');
			}
		}, $label);
	
	# Note that notify_id will be 0 if there was an error,
    # so we handle that in our destroy callback.	
	$label->{notify_id} = $notify_id;
	$label->{client} = $client;
	$label->signal_connect(destroy => sub {
			# Remove the notification callback when the widget monitoring
			# notifications is destroyed
			my $client = $_[0]->{client};
			my $notify_id = $_[0]->{notify_id};

			$client->notify_remove($notify_id) if $notify_id;
		});
	
	return $frame;
}

#
# Preferences dialog code. NOTE that the prefs dialog knows NOTHING
# about the existence of the main window; it is purely a way to fool
# with the GConf database. It never does something like change
# the main window directly; it ONLY changes GConf keys via
# GConfClient. This is _important_, because people may configure
# your app without using your preferences dialog.
#
# This is an instant-apply prefs dialog. For a complicated
# apply/revert/cancel dialog as in GNOME 1, see the
# complex-gconf-app.c example. But don't actually copy that example
# in GNOME 2, thanks. ;-) complex-gconf-app.c does show how
# to use GConfChangeSet.
#

sub prefs_dialog_destroyed
{
	my $dialog = shift;
	my $main_window = shift;

	$main_window->{prefs} = undef;
}

sub config_entry_commit
{
	my $entry = shift;

	my $client = $entry->{client};
	my $key    = $entry->{key};
	
	my $text = $entry->get_chars(0, -1);
	
	# Unset if the string is zero-length, otherwise set
	if ($text)
	{
		# show how to use the generic 'set' method, instead of get_string.
		# (ebassi)
		$client->set($key, { type => 'string', value => $text });
	}
	else
	{
		$client->unset($key);
	}
	
	# since we connect the "focus_out_event" to this callback, this return is
	# needed. (ebassi)
	return FALSE;
}

sub create_config_entry
{
	my $prefs_dialog = shift;
	my $client       = shift;
	my $config_key   = shift;
	my $has_focus    = shift || FALSE;

	my $hbox  = Gtk2::HBox->new(FALSE, 5);
	my $label = Gtk2::Label->new($config_key);
	my $entry = Gtk2::Entry->new;

	$hbox->pack_start($label, FALSE, FALSE, 0);
	$hbox->pack_end($entry, FALSE, FALSE, 0);

	# this will print an error via default error handler
    # if the key isn't set to a string
	my $s = $client->get_string($config_key);
	$entry->set_text($s) if $s;

	$entry->{client} = $client;
	$entry->{key} = $config_key;
	
	# Commit changes if the user focuses out, or hits enter; we don't
    # do this on "changed" since it'd probably be a bit too slow to
    # round-trip to the server on every "changed" signal.
	$entry->signal_connect(activate        => \&config_entry_commit);
	$entry->signal_connect(focus_out_event => \&config_entry_commit);
	
	# Set the entry insensitive if the key it edits isn't writable.
    # Technically, we should update this sensitivity if the key gets
    # a change notify, but that's probably overkill.
	$entry->set_sensitive($client->key_is_writable($config_key));

	$entry->grab_focus if $has_focus;
	
	return $hbox;
}


sub create_prefs_dialog
{
	my $parent = shift;
	my $client = shift;

	my $dialog = Gtk2::Dialog->new("basic-gconf-app Preferences",
								   $parent,
								   [ qw/destroy-with-parent/ ],
								   'gtk-close', 'accept');
	# destroy dialog on button press
	$dialog->signal_connect(response => sub { $_[0]->destroy });
	$dialog->set_default_response('accept');
	
	# resizing doesn't grow the entries anyhow
	$dialog->set_resizable(FALSE);

	my $vbox = Gtk2::VBox->new(FALSE, 5);
	$vbox->set_border_width(5);

	$dialog->vbox->pack_start($vbox, FALSE, FALSE, 0);

	my $entry;
	$entry = create_config_entry ($dialog, $client, "/apps/basic-gconf-app/foo", TRUE);
	$vbox->pack_start($entry, FALSE, FALSE, 0);

	$entry = create_config_entry ($dialog, $client, "/apps/basic-gconf-app/bar");
	$vbox->pack_start($entry, FALSE, FALSE, 0);

	$entry = create_config_entry ($dialog, $client, "/apps/basic-gconf-app/baz");
	$vbox->pack_start($entry, FALSE, FALSE, 0);

	$entry = create_config_entry ($dialog, $client, "/apps/basic-gconf-app/blah");
	$vbox->pack_start($entry, FALSE, FALSE, 0);
		
	return $dialog;
}


Makefile.PL
#
# $Id$
#

use 5.008;
use strict;
use warnings;
use Cwd;
use File::Spec;
use ExtUtils::MakeMaker;

# minimum required version of dependancies we need to build
our %build_reqs = (
	'perl-ExtUtils-Depends'   => '0.2',
	'perl-ExtUtils-PkgConfig' => '1.03',
	'perl-Glib'               => '1.120', # for Glib::MakeHelper
	'GConf'                   => '2.0.0',
);

# Writing a fake Makefile ensures that CPAN will pick up the correct
# dependencies and install them.
unless (eval "use ExtUtils::Depends '$build_reqs{'perl-ExtUtils-Depends'}';"
           . "use ExtUtils::PkgConfig '$build_reqs{'perl-ExtUtils-PkgConfig'}';"
	   . "use Glib::CodeGen;"
           # just seeing if Glib is available isn't enough, make sure
           # it's recent enough, too 
           . "use Glib '$build_reqs{'perl-Glib'}';"
           . "use Glib::MakeHelper;"
           . "1") {
   warn "$@\n";
   WriteMakefile(
         PREREQ_FATAL => 1,
         PREREQ_PM    => {
            'Glib'                => $build_reqs{'perl-Glib'},
            'ExtUtils::Depends'   => $build_reqs{'perl-ExtUtils-Depends'},
            'ExtUtils::PkgConfig' => $build_reqs{'perl-ExtUtils-PkgConfig'},
         },
   );
   exit 1; # not reached
}

mkdir 'build', 0777;

our %pkgcfg = ExtUtils::PkgConfig->find ('gconf-2.0 >= '.$build_reqs{GConf});


our @xs_files = <xs/*.xs>;
our %pm_files = (
    'GConf.pm' => '$(INST_LIBDIR)/GConf.pm',
);
our %pod_files = (
    'GConf.pm' => '$(INST_MAN3DIR)/Gnome2::GConf.$(MAN3EXT)',
    Glib::MakeHelper->do_pod_files (@xs_files),
);

ExtUtils::PkgConfig->write_version_macros ("build/gconfperl-version.h",
                                           'gconf-2.0' => 'GCONF');

#
# autogeneration
#
Glib::CodeGen->parse_maps ('gconfperl');
Glib::CodeGen->write_boot (ignore => '^Gnome2::GConf$');

# now we're ready to start creating the makefile.
# we need to use ExtUtils::Depends to get relevant information out of
# the Glib extension, and to save config information for other modules which
# will chain from this one.
our $gconf = ExtUtils::Depends->new ('Gnome2::GConf', 'Glib');
$gconf->set_inc ($pkgcfg{cflags} . ' -I./build ');
$gconf->set_libs ($pkgcfg{libs});
$gconf->add_xs (@xs_files);
$gconf->add_pm (%pm_files);
my $cwd = cwd();
$gconf->add_typemaps (map { File::Spec->catfile($cwd, $_) }
                      'build/gconfperl.typemap', 'gconf.typemap');
$gconf->install (qw(gconfperl.h
		build/gconfperl-autogen.h
		build/gconfperl-version.h
		doctypes));
$gconf->save_config ('build/IFiles.pm');

WriteMakefile(
    NAME          => 'Gnome2::GConf',
    VERSION_FROM  => 'GConf.pm', # finds $VERSION
    PREREQ_PM     => {}, # e.g., Module::Name => 1.1
    ABSTRACT_FROM => 'GConf.pm', # retrieve abstract from module
    XSPROTOARG    => '-noprototypes',
    MAN3PODS	  => \%pod_files,
    $gconf->get_makefile_vars,
);

#print "
#WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING
#
#This is an unstable development release of Gnome2::GConf.  The API is
#not frozen and things are subject to change at any time.  Report any
#bugs to gtk-perl-list AT gnome DOT org as soon as possible.
#Please use the stable 1.020 version for important work.
#
#WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING
#
#";
			

sub MY::postamble
{
	return Glib::MakeHelper->postamble_clean ()
	     . Glib::MakeHelper->postamble_docs_full (
		DEPENDS => $gconf,
		DOCTYPES => 'doctypes',
		COPYRIGHT_FROM => 'copyright.pod',
	       )
	     . Glib::MakeHelper->postamble_rpms (
		'GCONF' => $build_reqs{'GConf'},
		'PERL_EXTUTILS_DEPENDS' => 
			$build_reqs{'perl-ExtUtils-Depends'},
		'PERL_EXTUTILS_PKGCONFIG' => 
			$build_reqs{'perl-ExtUtils-PkgConfig'},
		'PERL_GLIB' => $build_reqs{'perl-Glib'},
	       );
}
